Technical review: Document customizable select listboxes (#43352)
mdn/content: files/en-us/learn_web_development/extensions/forms/customizable_select_listboxes/index.md

---
title: Customizable select listboxes
short-title: Customizable listboxes
slug: Learn_web_development/Extensions/Forms/Customizable_select_listboxes
page-type: learn-module-chapter
sidebar: learnsidebar
---

{{PreviousMenuNext("Learn_web_development/Extensions/Forms/Customizable_select", "Learn_web_development/Extensions/Forms/UI_pseudo-classes", "Learn_web_development/Extensions/Forms")}}

This article follows on from the previous one, looking at how to style customizable listbox {{htmlelement("select")}} elements.

## Dropdown selects versus listboxes

> When we talk about "dropdown" `<select>` elements, we are talking about the controls rendered for HTML like `<select>`. These feature a select button that, when pressed, shows a dropdown picker from which you can select different options. "Listbox" select elements on the other hand are controls rendered for HTML like `<select multiple>` or `<select size="3">`. These feature a box that shows multiple options at once, from which you can select one or multiple options.

Reviewer: The first sentence needs work. It reads like select is a select. I understand you're trying to differentiate a bare select with one that has attributes, but this doesn't work. The comparative example below works really well.

  <select id="pet-select" multiple>
    <div class="wrapper">
      <option value="cat">Cat</option>
      <option value="dog">Dog</option>
      <option value="chicken">Chicken</option>
      <option value="fish">Fish</option>
      <option value="hamster">Hamster</option>
      <option value="gerbil">Gerbil</option>
      <option value="guinea">Guinea pig</option>
    </div>
  </select>
</p>
```

In the CSS, we start by setting the containing {{htmlelement("p")}} element's {{cssxref("width")}} and {{cssxref("margin")}} so that the demo will be centered horizontally in the viewport and take up most of the width. We then size the `<select>` to take up the full width of its parent and only be as tall as the `<option>` elements. The `.wrapper` `<div>` is given a {{cssxref("display")}} value of `flex`, causing the `<option>` elements to be laid out horizontally in a row; we then set its `width` so that it is always as wide as the `<option>` elements.

```css live-sample___horizontal-listbox
p {
  width: 90%;
  margin: 0 auto;
}

select {
  width: 100%;
  height: fit-content;
}

.wrapper {
  display: flex;
  width: fit-content;
}
```

> Next, we give the `<option>` elements some extra padding to space them out horizontally, and a {{cssxref("position")}} value of relative so we can position their descendents relative to them.

Reviewer: Preferred spelling I understand
```suggestion
Next, we give the `<option>` elements some extra padding to space them out horizontally, and a {{cssxref("position")}} value of relative so we can position their descendants relative to them.
```

```js live-sample___complex-listbox
select.addEventListener("change", () => {
  const allCurrentValues = Array.from(select.options).map(
    (option) => option.value,
  );
  const currentSelectedValues = Array.from(select.selectedOptions).map(
    (option) => option.value,
  );

  contacts.forEach((contact) => {
    if (allCurrentValues.includes(contact.name)) {
      contact.selected = currentSelectedValues.includes(contact.name);
    }
  });
});
```

Finally, we run the `populateOptions()` function, passing it the `contacts` array, so that on page load the full list of contacts is displayed.

```js live-sample___complex-listbox
populateOptions(contacts);
```

### Result

The example renders like so:

> {{EmbedLiveSample("complex-listbox", "100%", "360px")}}

Reviewer: Does not render on PR or locally.

<img width="1147" height="603" alt="Image" src="https://github.com/user-attachments/assets/4af8a714-90dd-4763-ab72-692d71599b90" />

Reviewer: Sometimes things glitch. Working great now.

---
title: Customizable select listboxes
short-title: Customizable listboxes
slug: Learn_web_development/Extensions/Forms/Customizable_select_listboxes
page-type: learn-module-chapter
sidebar: learnsidebar
---

{{PreviousMenuNext("Learn_web_development/Extensions/Forms/Customizable_select", "Learn_web_development/Extensions/Forms/UI_pseudo-classes", "Learn_web_development/Extensions/Forms")}}

> This article follows on from the previous one, looking at how to style customizable listbox {{htmlelement("select")}} elements.

Reviewer: Does this read like maybe the previous one was covering "looking a how"? Maybe

```suggestion
This article follows on from the previous one, and looks at how to style customizable listbox {{htmlelement("select")}} elements.
```

One of the major advantages of customizable `<select>` listboxes over "classic" select listboxes is that you can fully style all parts of the control, and you can include a much wider variety of child elements inside them, which means greater flexibility in terms of design and functionality.

## Dropdown selects versus listboxes

When we talk about "dropdown" `<select>` elements, we are talking about controls featuring a select button that, when pressed, shows a dropdown picker from which you can select an option. These are specified using basic HTML such as `<select>`.

> "Listbox" `<select>` elements on the other hand are controls featuring a box that shows multiple options at once, from which you can select one or multiple options. You opt into rendering a "listbox" select by specifying the `multiple` attribute (to allow multiple selections) and/or a `size` value of more than `1`. For example, `<select multiple>` or `<select size="3">`.

Reviewer: This might be annoying, since you already touched it, but it feels like listbox should be the focus

```suggestion
## Select listboxes vs dropdown selects

In the previous article we talked about "dropdown" `<select>` elements, which are controls that have a button that, when pressed, shows a dropdown picker from which you can select an option. These are specified using basic HTML such as `<select>`.

"Listbox" `<select>` elements by contrast are controls featuring a box that shows multiple options at once, from which you can select one or multiple options. You opt into rendering a "listbox" select by specifying the `multiple` attribute (to allow multiple selections) and/or a `size` value of more than `1`. For example, `<select multiple>` or `<select size="3">`.
```

The following live example illustrates the difference:

```html hidden live-sample___select-comparison
<form>
  <p>
    <label for="pet-select">Select pet dropdown:</label><br />
    <select id="pet-select">
      <option value="cat">Cat</option>
      <option value="dog">Dog</option>
      <option value="chicken">Chicken</option>
      <option value="fish">Fish</option>
      <option value="Hamster">Hamster</option>
    </select>
  </p>
  <p>
    <label for="pet-select2">Select pets listbox:</label><br />
    <select id="pet-select2" multiple>
      <option value="cat">Cat</option>
      <option value="dog">Dog</option>
      <option value="chicken">Chicken</option>
      <option value="fish">Fish</option>

>       <option value="Hamster">Hamster</option>

Reviewer: ```suggestion
      <option value="hamster">Hamster</option>
```

```html live-sample___basic-listbox live-sample___expanding-listbox
<p>
  <label for="pet-select">Select pets:</label><br />
  <select id="pet-select" multiple>
    <option value="cat">Cat</option>
    <option value="dog">Dog</option>
    <option value="chicken">Chicken</option>
    <option value="fish">Fish</option>
    <option value="Hamster">Hamster</option>
  </select>
</p>
```

There is nothing remarkable here. Note that we are rendering our listbox using `<select multiple>` rather than `<select size="3">`. The only difference is that we can select multiple options rather than a single option. The styling works in exactly the same way.

We begin our styling by opting the `<select>` into custom styling with an {{cssxref("appearance")}} value of `base-select`:

```css hidden live-sample___basic-listbox live-sample___expanding-listbox live-sample___horizontal-listbox
* {
  box-sizing: border-box;
}

html {
  font-family: Arial, Helvetica, sans-serif;
}
```

```css live-sample___basic-listbox live-sample___expanding-listbox live-sample___horizontal-listbox
select {
  appearance: base-select;
}
```

> This enables us to style our {{htmlelement("select")}} and {{htmlelement("option")}} elements however we want. Our basic styles look like this:

Reviewer: Feels like we could separate this a little. Up to you though.
```suggestion
With that done we are now able to style our {{htmlelement("select")}} and {{htmlelement("option")}} elements however we want.

Our basic styles look like this:
```
